app/scripts/controllers/network/network-controller.js

```diff
@@ -476,4 +501,113 @@ export default class NetworkController extends EventEmitter {
   }
+
+  /**
+   * Network Configuration management functions
+   */
+
+  /**
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
+   *
+   * @param {NetworkConfiguration} networkConfiguration - The network configuration to add or, if rpcUrl matches an existing entry, to modify.
+   * @param {object} options
+   * @param {boolean} options.setActive - An option to set the newly added networkConfiguration as the active provider.
+   * @param {string} options.referrer - The site from which the call originated, or 'metamask' for internal calls - used for event metrics.
+   * @param {string} options.source - Where the upsertNetwork event originated (i.e. from a dapp or from the network form)- used for event metrics.
+   * @returns {string} id for the added or updated network configuration
+   */
+  upsertNetworkConfiguration(
+    { rpcUrl, chainId, ticker, nickname, rpcPrefs },
+    { setActive = false, referrer, source } = {},
+  ) {
+    assert.ok(
+      isPrefixedFormattedHexString(chainId),
+      `Invalid chain ID "${chainId}": invalid hex string.`,
+    );
+    assert.ok(
+      isSafeChainId(parseInt(chainId, 16)),
+      `Invalid chain ID "${chainId}": numerical value greater than max safe value.`,
+    );
+
+    if (!rpcUrl) {
+      throw new Error(
+        'An rpcUrl is required to add or update networkConfiguration',
+      );
+    }
+
+    try {
+      // eslint-disable-next-line no-new
+      new URL(rpcUrl);
+    } catch (e) {
+      if (e.message.includes('Invalid URL')) {
+        throw new Error('rpcUrl must be a valid URL');
+      }
+    }
+
+    if (!ticker) {
+      throw new Error(
+        'A ticker is required to add or update networkConfiguration',
+      );
+    }
+
+    const networkConfigurations = this.networkConfigurationsStore.getState();
+    const newNetworkConfiguration = {
+      rpcUrl,
+      chainId,
+      ticker,
+      nickname,
+      rpcPrefs,
+    };
+
+    const oldNetworkConfigurationId = Object.values(networkConfigurations).find(
+      (networkConfiguration) =>
+        networkConfiguration.rpcUrl?.toLowerCase() === rpcUrl?.toLowerCase(),
+    )?.id;
+
+    const newNetworkConfigurationId = oldNetworkConfigurationId || random();
+    this.networkConfigurationsStore.updateState({
+      ...networkConfigurations,
+      [newNetworkConfigurationId]: {
+        ...newNetworkConfiguration,
+        id: newNetworkConfigurationId,
+      },
+    });
+
+    if (!oldNetworkConfigurationId) {
+      this._trackMetaMetricsEvent({
+        event: 'Custom Network Added',
+        category: EVENT.CATEGORIES.NETWORK,
+        referrer: {
+          url: referrer,
+        },
+        properties: {
+          chain_id: chainId,
+          symbol: ticker,
+          source: Object.values(EVENT.SOURCE.NETWORK).includes(source)
+            ? source
+            : EVENT.SOURCE.TRANSACTION.DAPP,
```

Comment:
Thoughts on `source` directly here? We ought to be able to set this to `EVENT.SOURCE.TRANSACTION.DAPP` explicitly in the `wallet_addEthereumChain` handler. That would simplify things somewhat.

```suggestion
          source,
```

Reply:
Oh yes this was an oversight!

Reply:
fixed here: https://github.com/MetaMask/metamask-extension/pull/17421/commits/35baacb51df95929f05d2082f3c837689fb9a146